tee-worker/litentry/core/data-providers/src/graphql.rs

```diff
@@ -0,0 +1,191 @@
+	pub token_address: String,
+	pub mini_balance: f64,
+}
+
+impl VerifiedCredentialsIsHodlerIn {
+	pub fn new(
+		addresses: Vec<String>,
+		from_date: String,
+		network: VerifiedCredentialsNetwork,
+		token_address: String,
+		mini_balance: f64,
+	) -> Self {
+		VerifiedCredentialsIsHodlerIn { addresses, from_date, network, token_address, mini_balance }
+	}
+
+	pub fn conv_to_string(&self) -> String {
+		let mut flat = "addresses:[".to_string();
+		for addr in self.addresses.iter() {
+			flat += &format!("\"{}\",", addr);
+		}
+		flat += "],";
+		flat += &format!("fromDate:\"{}\",", self.from_date.clone());
+		match &self.network {
+			VerifiedCredentialsNetwork::Litentry => flat += "network:litentry",
+			VerifiedCredentialsNetwork::Litmus => flat += "network:litmus",
+			VerifiedCredentialsNetwork::Polkadot => flat += "network:polkadot",
+			VerifiedCredentialsNetwork::Kusama => flat += "network:kusama",
+			VerifiedCredentialsNetwork::Khala => flat += "network:khala",
+			VerifiedCredentialsNetwork::Ethereum => flat += "network:ethereum",
+		}
+		if self.token_address.is_empty().not() {
+			flat += &format!(",tokenAddress:\"{}\"", &self.token_address.clone());
+		}
+		flat += &format!(",minimumBalance:{:?}", self.mini_balance.clone());
+		flat
+	}
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct QLResponse {
+	#[serde(flatten)]
+	// data: HashMap<String, serde_json::Value>,
+	data: serde_json::Value,
+}
+impl RestPath<String> for QLResponse {
+	fn get_path(path: String) -> core::result::Result<String, HttpError> {
+		Ok(path)
+	}
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "PascalCase")]
+pub struct IsHodlerOut {
+	pub verified_credentials_is_hodler: Vec<IsHodlerOutStruct>,
+}
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct IsHodlerOutStruct {
+	pub address: String,
+	pub is_hodler: bool,
+}
+
+impl GraphQLClient {
+	pub fn new() -> Self {
+		let mut headers = Headers::new();
+		headers.insert(CONNECTION.as_str(), "close");
+		headers.insert(
+			AUTHORIZATION.as_str(),
+			G_DATA_PROVIDERS.read().unwrap().graphql_auth_key.clone().as_str(),
+		);
+		let client =
+			build_client(G_DATA_PROVIDERS.read().unwrap().graphql_url.clone().as_str(), headers);
+		GraphQLClient { client }
+	}
+
+	pub fn verified_credentials_is_hodler(
+		&mut self,
+		credentials: VerifiedCredentialsIsHodlerIn,
+	) -> Result<IsHodlerOut, Error> {
+		// FIXME: for the moment, the `path` is partially hard-code here.
+		let path = "latest/graphql?query=query{VerifiedCredentialsIsHodler(".to_string()
+			+ &credentials.conv_to_string()
+			+ "){isHodler, address}}";
+
+		let response = self
+			.client
+			.get_with::<String, QLResponse>(path, vec![].as_slice())
+			.map_err(|e| Error::RequestError(format!("{:?}", e)))?;
+
+		if let Some(value) = response.data.get("data") {
+			let is_hodler_out: IsHodlerOut = serde_json::from_value(value.clone()).unwrap();
```

Comment:
need to make sure or add some comment that `unwrap()` is much safe to use here, or return an Invalid data error.